### PR TITLE
Add 3D extent to providers (2nd stage)

### PR DESCRIPTION
**Pull request by benoitdm-oslandia**
This PR follows #54941. We have add 3D extent implementation for these providers:

* QgsOgrProvider
* QgsPostgresProvider
* QgsSpatiaLiteProvider
* QgsDelimitedTextProvider
* QgsPointCloudLayer 

These implementations could be discussed and the pointCloud one is a draft.

Funded by CEA/DAM @renardf

cc @ptitjano, @lbartoletti, @troopa81, @rouault, @uclaros

### CODE REVIEW

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -1117,16 +1135,56 @@ unsigned char *QgsOgrProvider::getGeometryPointer( OGRFeatureH fet )
...
+    {
+      OGREnvelope3D env;
+      OGR_G_GetEnvelope3D( g, &env );
+
+      extent->MinX = std::min( extent->MinX, env.MinX );
+      extent->MinY = std::min( extent->MinY, env.MinY );
+      extent->MaxX = std::max( extent->MaxX, env.MaxX );
+      extent->MaxY = std::max( extent->MaxY, env.MaxY );
+      extent->MinZ = std::min( extent->MinZ, env.MinZ );
+      extent->MaxZ = std::max( extent->MaxZ, env.MaxZ );
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrprovider.cpp:1160]: if there's a mix of 2D/3D objects in the layer, MinZ/MaxZ will be 0 on 2D objects

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 2]
```diff
@@ -1136,55 +1194,95 @@ QgsRectangle QgsOgrProvider::extent() const
...
+            QgsDebugMsgLevel( QStringLiteral( "WITH OLCFastGetExtent AND WITH geomCol: %1" ).arg( geomCol ), 3 );
+
+            mOgrLayer->GetExtent( mExtent.get(), true );
+            QByteArray geomColQuoted = quotedIdentifier( geomCol );
+            QByteArray sql = "SELECT MIN(ST_MinZ("
+                             + geomColQuoted
+                             + ")), MAX(ST_MaxZ("
+                             + geomColQuoted
+                             + ")) FROM "
+                             + quotedIdentifier( mOgrLayer->name() );
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrprovider.cpp:1235]: this will only work with PostGIS, GPKG and Spatialite datasets. You probably want to run the ExecuteSQL() between CPLPushErrorHandler(CPLQuietErrorHandler) / CPLPopErrorHandler() to avoid error messages on datasets that don't support that

File: tests/src/python/test_provider_ogr.py [hunk 3]
```diff
@@ -3497,6 +3498,41 @@ def testGeoJsonMapType(self):
...
+        self.assertTrue(vl.featureCount(), 9)
+        self.assertEqual(vl.extent(), QgsRectangle(0, -3, 16, 5))
+        self.assertEqual(vl.extent3D(), QgsBox3D(0, -3, float('nan'), 16, 5, float('nan')))
+        del vl
+
+        os.unlink(datasource_2d)
+        self.assertFalse(os.path.exists(datasource_2d))
+
+        # 3D points
+        datasource_3d = os.path.join(self.basetestpath, 'testExtent3D.csv')
```
Comment (rouault) [inline tests/src/python/test_provider_ogr.py:3520]: you'll probably also want to do a similar test on GeoPackage to the the "SELECT min(ST_MinZ(...) ..." code path

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 4]
```diff
@@ -464,6 +464,24 @@ QgsOgrProvider::QgsOgrProvider( QString const &uri, const ProviderOptions &optio
   }
   setNativeTypes( nativeTypes );
 
+  if ( mOgrOrigLayer )
+  {
+    bool is3D = true;
+    gdal::ogr_feature_unique_ptr f;
+
+    mOgrOrigLayer->ResetReading();
```
Comment (nyalldawson) [inline src/core/providers/ogr/qgsogrprovider.cpp:472]: Why are we sniffing a feature here instead of just checking the wkb type of the layer?

File: src/providers/postgres/qgspostgresprovider.cpp [hunk 5]
```diff
@@ -247,6 +247,26 @@ QgsPostgresProvider::QgsPostgresProvider( QString const &uri, const ProviderOpti
     }
   }
 
+  // check if geometry is 3D or not
+  if ( !mGeometryColumn.isEmpty() && !mTableName.isEmpty() )
+  {
+    const QString selectQuery = QStringLiteral( R"SQL(
```
Comment (nyalldawson) [inline src/providers/postgres/qgspostgresprovider.cpp:253]: Again, why not just rely on the wkb type here?

File: src/core/providers/ogr/qgsogrprovider.cpp (hunk 4 above)
Comment (rouault) [reply src/core/providers/ogr/qgsogrprovider.cpp:472]: > Why are we sniffing a feature here instead of just checking the wkb type of the layer?

That was my initial reaction, but there are a number of data sources where the typing is loose. That said the strategy could be :
1. look at the wkb type of the layer: if it is 3D, then note it and stop
2. otherwise sniff a feature

File: src/core/providers/ogr/qgsogrprovider.cpp (hunk 1 above)
Comment (benoitdm-oslandia) [reply src/core/providers/ogr/qgsogrprovider.cpp:1160]: Is there a reason to obtain 0 for non existent values? Do you think gdal can instead return nan?

May be to solve this, we can retrieve the 2D/3D state in each row?

Comment (benoitdm-oslandia) [reply src/core/providers/ogr/qgsogrprovider.cpp:1160]: I will try using `OGR_G_Is3D`

File: src/core/providers/ogr/qgsogrprovider.cpp (hunk 4 above)
Comment (benoitdm-oslandia) [reply src/core/providers/ogr/qgsogrprovider.cpp:472]: fixed!

File: src/providers/postgres/qgspostgresprovider.cpp (hunk 5 above)
Comment (benoitdm-oslandia) [reply src/providers/postgres/qgspostgresprovider.cpp:253]: fixed

File: src/core/providers/ogr/qgsogrprovider.cpp (hunk 1 above)
Comment (rouault) [reply src/core/providers/ogr/qgsogrprovider.cpp]: > Is there a reason to obtain 0 for non existent values? Do you think gdal can instead return nan?

I'm not sure of the reason besides being what is implemented... Note that NaN is also tricky to use with std::min() / std::max() (cf https://stackoverflow.com/questions/72296440/how-to-prevent-stdmin-and-max-to-return-nan-if-the-first-element-of-the-array)

> May be to solve this, we can retrieve the 2D/3D state in each row? I will try using OGR_G_Is3D

yes, that's what I had in mind

File: src/core/providers/ogr/qgsogrprovider.cpp (hunk 2 above)
Comment (rouault) [reply src/core/providers/ogr/qgsogrprovider.cpp:1242]: > The check `if ( l ) {...}` will be enough?

yes

File: tests/src/python/test_provider_ogr.py (hunk 3 above)
Comment (benoitdm-oslandia) [reply tests/src/python/test_provider_ogr.py:3521]: done!

File: src/providers/postgres/qgspostgresprovider.cpp (hunk 5 above)
Comment (benoitdm-oslandia) [reply src/providers/postgres/qgspostgresprovider.cpp:257]: We will reduce readability because the schemas/tables are not the same. This request is made once, the impact will be low.

Comment (nyalldawson) [reply src/providers/postgres/qgspostgresprovider.cpp:257]: I respectfully disagree -- see eg all the work done by https://github.com/qgis/QGIS/pulls?q=is%3Apr+is%3Aclosed+author%3Amorriganr focused on removing every possible extraneous query. That's driven by real world requirements.

Comment (benoitdm-oslandia) [reply src/providers/postgres/qgspostgresprovider.cpp:257]: Ok but I don't know how to properly do this merged query. Maybe @MorriganR could help us find a solution (please :grin:)?

**Comment by benoitdm-oslandia** [comment]
I just fix a regression in my spatialite. Tests should pass now!

**Comment by benoitdm-oslandia** [comment]
@nyalldawson @rouault Tests are ok expect one with bad remote server access.

File: src/providers/postgres/qgspostgresprovider.cpp (hunk 5 above)
Comment (nyalldawson) [reply src/providers/postgres/qgspostgresprovider.cpp:257]: Can we drop the postgres changes for now then, and implement in a specific follow up PR? I don't want to see speed regressions introduced here.

Comment (benoitdm-oslandia) [reply src/providers/postgres/qgspostgresprovider.cpp:257]: Yes it is the best solution for now

**Review by nyalldawson — APPROVED** [review]

**Comment by benoitdm-oslandia** [comment]
rebase from master...

**Comment by uclaros** [comment]
Not a proper review, but you've left a couple of TODO comments there.
TBH, I find it quite hard to follow the review process when force-pushing...

**Review by rouault — APPROVED** [review]

**Comment by rouault** [comment]
@benoitdm-oslandia Now that https://github.com/OSGeo/gdal/commit/8655f56595ca4c199069d6654ded11503d4f9802 has landed into GDAL master, it could be good that you have a GDAL >= 3.9.0 code path that uses OGR_L_GetExtent3D()

**Comment by rouault** [comment]
> Does it work for all driver?

yes, there's a base implementation that will be the fallback for all drivers does what you implemented: iterating over all features. For shapefile, there's a specialized implementation that just gets the extent from the shapefile header.  GeoPackage and PostGIS specialized implementations are also planned.

**Comment by nyalldawson** [comment]
@benoitdm-oslandia the test failure here ISN'T a false positive -- that test needs updating to match these changes

**Comment by benoitdm-oslandia** [comment]
@nyalldawson my bad! Still works to do!

**Comment by benoitdm-oslandia** [comment]
@rouault does your implementation handle the `mSubsetString` or the new `OGR_L_GetExtent3D` function will only replace the code from line 1230 to 1263 (main call to OGR_L_GetExtent)? will it handle the case from line 1211 to 1224 (ie. WITHOUT OLCFastGetExtent)?

**Comment by rouault** [comment]
I'm CC'ing @elpaso who is implementing the functionality

> does your implementation handle the `mSubsetString` or the new `OGR_L_GetExtent3D` function will only replace the code from line 1230 to 1263 (main call to OGR_L_GetExtent)?

As far as I can see, that will be similar to GetExtent(). It is driver dependent whether they take into account or not filters. The generic (slow) implementation and as far as I can see the GeoPackage do, but the shapefile one does not. We could potentially make the shapefile implementation fallback to the generic (slow) implementation if there's a filter, and if taking it into account is desired.

> will it handle the case from line 1211 to 1224 (ie. WITHOUT OLCFastGetExtent)?

If the question is: will the OAPIF driver return a dummy extent when GetExtent3D() is called on it. The answer is: no, it won't, it will just go through the generic implementation.

**Comment by benoitdm-oslandia** [comment]
@rouault I add a fix without checking the tests. Does it look ok to you?

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 6]
```diff
@@ -1222,72 +1222,42 @@ QgsBox3D QgsOgrProvider::extent3D() const
...
       }
-      else
+      else if ( mSubsetString.isEmpty() ) // TODO: find a way to optimize when mSubsetString is defined
       {
-        if ( is3D && mSubsetString.isEmpty() ) // TODO: find a way to use mSubsetString in the request
+        QgsDebugMsgLevel( QStringLiteral( "WITH OLCFastGetExtent" ), 3 );
+        QRecursiveMutex *mutex = nullptr;
+        OGRLayerH layer = mOgrLayer->getHandleAndMutex( mutex );
+        const char *geomCol = OGR_L_GetGeometryColumn( layer );
+        if ( geomCol != NULL && strlen( geomCol ) > 0 )
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrprovider.cpp:1233]: you don't need that test anymore as you don't write SQL. You could just test OGR_L_GetGeomType() != wkbNone

File: src/core/providers/ogr/qgsogrproviderutils.cpp [hunk 7]
```diff
@@ -3106,6 +3106,24 @@ OGRErr QgsOgrLayer::GetExtent( OGREnvelope *psExtent, bool bForce )
   return OGR_L_GetExtent( hLayer, psExtent, bForce );
 }
 
+OGRErr QgsOgrLayer::GetExtent3D( OGREnvelope3D *psExtent3D, bool bForce )
+{
+  QMutexLocker locker( &ds->mutex );
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(3,9,0)
+  return OGR_L_GetExtent3D( hLayer, psExtent3D, bForce );
+#else
+  OGREnvelope psExtent;
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrproviderutils.cpp:3115]: you should probably fallback to compute3DExtentSlowly() to get same results in QGIS whatever the GDAL version

**Comment by rouault** [comment]
> I add a fix without checking the tests

What do you mean?

File: src/core/providers/ogr/qgsogrprovider.cpp (hunk 6 above)
Comment (benoitdm-oslandia) [reply src/core/providers/ogr/qgsogrprovider.cpp:1233]: ok fixed

**Comment by benoitdm-oslandia** [comment]
rebase from master, handle future gdal 3.9 extend3D function and fix tests

File: src/core/providers/ogr/qgsogrproviderutils.cpp (hunk 7 above)
Comment (benoitdm-oslandia) [reply src/core/providers/ogr/qgsogrproviderutils.cpp]: done

**Comment by benoitdm-oslandia** [comment]
fix authors...

**Comment by benoitdm-oslandia** [comment]
@rouault I have a question about gdal. 

Context: with `orginfo -al tests/testdata/points.shp` I obtain an extent with xmin=-118.8 but there is no point in this file with such coordinate. 

Question: does `ogrinfo` re-compute the extent each time or does it read it from the metadata of the shapefile? Is there a way to force the re-computation of the extent?

**Comment by nyalldawson** [comment]
@benoitdm-oslandia regarding [695feac](https://github.com/qgis/QGIS/pull/55320/commits/695feacacdaac0398eecd78aab8a3a9b8be11b1a) -- what's triggering the new request? I don't think we should be calculating the 3d extent on layer load there, rather it should be deferred till it's actually required.

**Comment by benoitdm-oslandia** [comment]
@nyalldawson can you give the right permission on GitHub to restart a failed job please?

**Comment by benoitdm-oslandia** [comment]
> @benoitdm-oslandia regarding [695feac](https://github.com/qgis/QGIS/pull/55320/commits/695feacacdaac0398eecd78aab8a3a9b8be11b1a) -- what's triggering the new request? I don't think we should be calculating the 3d extent on layer load there, rather it should be deferred till it's actually required.

@nyalldawson, we are just reading the first feature to compute if the layer is 2D or 3D. We do not calculate the extent on layer load.

All lights are green and I am all done! Tell me if you have some point to discuss!

File: src/core/providers/ogr/qgsogrproviderutils.cpp [hunk 8]
```diff
@@ -3106,6 +3106,92 @@ OGRErr QgsOgrLayer::GetExtent( OGREnvelope *psExtent, bool bForce )
   return OGR_L_GetExtent( hLayer, psExtent, bForce );
 }
 
+OGRErr QgsOgrLayer::GetExtent3D( OGREnvelope3D *psExtent3D, bool bForce )
+{
+  QMutexLocker locker( &ds->mutex );
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(3,9,0)
+  return OGR_L_GetExtent3D( hLayer, psExtent3D, bForce );
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrproviderutils.cpp:3113]:
```suggestion
  return OGR_L_GetExtent3D( hLayer, /* iGeomField = */ 0, psExtent3D, bForce );
```

**Comment by rouault** [comment]
> Are you ok if we merge?

can you give it a try with GDAL master locally, and check the test_provider_ogr*.py tests pass  ?

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 9]
```diff
@@ -1117,16 +1139,32 @@ unsigned char *QgsOgrProvider::getGeometryPointer( OGRFeatureH fet )
 
 
 QgsRectangle QgsOgrProvider::extent() const
+{
+  return extent3D().toRectangle();
+}
+
+// ================================== TODO: NOT CONST AT ALL!!!!
```
Comment (uclaros) [inline src/core/providers/ogr/qgsogrprovider.cpp:1146]: todo?

Comment (benoitdm-oslandia) [reply src/core/providers/ogr/qgsogrprovider.cpp:1146]: fixed! the extent member is mutable!

**Comment by benoitdm-oslandia** [comment]
@rouault 

> can you give it a try with GDAL master locally, and check the test_provider_ogr*.py tests pass ?

The tests does not pass when the subset (attribute filter) is defined... Still checking why!

**Comment by benoitdm-oslandia** [comment]
@rouault here is the path taken from qgis by asking `getExtent3D` on a 2D gpkg: 
1. `QgsOgrProvider::extent3D`
2. `QgsOgrLayer::GetExtent3D`
3. `OGR_L_GetExtent3D`
4. `OGRGeoPackageTableLayer::GetExtent3D` ==> as we are 2D we switch to 2D version
5. `OGRLayer::GetExtent`
6. `OGRGeoPackageTableLayer::GetExtent` ==> does not handle subset (aka attribute filter) ==> test failure

As we have said lately, `OGRGeoPackageTableLayer::GetExtent` will not handle the subset to avoid to break the existing behaviour. So I will add a check (qgis side) if the layer is 2D and if there is a subset defined, I will use the slow computed version. Can I suppose `QgsOgrLayer::GetExtent3D` implementations will handle the subset?

**Comment by rouault** [comment]
> As we have said lately, `OGRGeoPackageTableLayer::GetExtent` will not handle the subset to avoid to break the existing behaviour.

should now be fixed in GDAL master per https://github.com/OSGeo/gdal/commit/d26bb2861584d2588adc96478397417f4657df57

**Comment by benoitdm-oslandia** [comment]
@rouault We pushed at the same time! I re-test with your new version.

**Comment by benoitdm-oslandia** [comment]
Good for me!

**Comment by benoitdm-oslandia** [comment]
Add postgresql implementation

**Comment by benoitdm-oslandia** [comment]
@nyalldawson @rouault everything is ok for you? Can we merge?

**Comment by rouault** [comment]
OK for me (mostly reviewed the OGR part)

**Comment by benoitdm-oslandia** [comment]
@nyalldawson as this PR has been approved, can you merge it please? Thanks!

File: src/providers/postgres/qgspostgresprovider.cpp [hunk 10]
```diff
@@ -1082,14 +1088,24 @@ bool QgsPostgresProvider::loadFields()
...
     QString fieldName = result.PQfname( i );
+    Oid tableoid = result.PQftable( i );
+    int attnum = result.PQftablecol( i );
+    QString formattedFieldType = fmtFieldTypeMap[tableoid][attnum];
     if ( fieldName == mGeometryColumn )
+    {
+      if ( mDetectedGeomType == Qgis::WkbType::Unknown )
+      {
+        // rare case, elevation can not be computed via mDetectedGeomType, we parse the formattedFieldType and look for Z data
+        const thread_local QRegularExpression re( QRegularExpression::anchoredPattern( "geometry\\(\\w+ZM?(,\\d+)?\\)" ), QRegularExpression::CaseInsensitiveOption );
```
Comment (lbartoletti) [inline src/providers/postgres/qgspostgresprovider.cpp:1099]:
```suggestion
        const thread_local QRegularExpression re( QRegularExpression::anchoredPattern( "geometry\\(\\w+\s*Z?M?(,\\d+)?\\)" ), QRegularExpression::CaseInsensitiveOption );
```

Suggested changes:
- \s* : Sometimes WKT-like pattern is "POINT ZM" and not "POINTZM"
- Z?: Add a "?" here. Because, without "?", you only look for "Z" type. But, maybe it's normal?

File: src/providers/postgres/raster/qgspostgresrastershareddata.cpp [hunk 11]
```diff
@@ -304,7 +304,7 @@ bool QgsPostgresRasterSharedData::fetchTilesIndex( const QgsGeometry &requestPol
   // Include actual bounds
   if ( !overallExtent.isNull() )
   {
-    mLoadedIndexBounds[ cacheKey ] = requestPolygon.combine( QgsGeometry::fromRect( overallExtent ) );
+    mLoadedIndexBounds[ cacheKey ] = requestPolygon.combine( QgsGeometry::fromWkt( overallExtent.asWktPolygon() ) );
```
Comment (lbartoletti) [inline src/providers/postgres/raster/qgspostgresrastershareddata.cpp:307]: Why use `fromWkt`? You can “lose” precision during the conversion double->str->double

File: tests/src/python/test_qgsdelimitedtextprovider.py [hunk 12]
```diff
@@ -878,6 +879,8 @@ def test_044_ZM(self):
         assert vl.isValid(), f"{basetestfile} is invalid"
         assert vl.wkbType() == QgsWkbTypes.PointZM, "wrong wkb type, should be PointZM"
         assert vl.getFeature(2).geometry().asWkt() == "PointZM (-71.12300000000000466 78.23000000000000398 1 2)", "wrong PointZM geometry"
+        self.assertEqual(vl.extent(), QgsRectangle(-71.12300000000000466, 66.32999999999999829, -65.31999999999999318, 78.29999999999999716))
```
Comment (lbartoletti) [inline tests/src/python/test_qgsdelimitedtextprovider.py:882]:
```suggestion
        self.assertEqual(vl.extent(), QgsRectangle(-71.12300000000000466, 66.32999999999999829, -65.31999999999999318, 78.29999999999999716))
```

Theses tests will be flaky on different machines due to precision. Hope, you can use https://github.com/qgis/QGIS/pull/55648

File: src/providers/postgres/raster/qgspostgresrastershareddata.cpp (hunk 11 above)
Comment (ptitjano) [reply src/providers/postgres/raster/qgspostgresrastershareddata.cpp]: Thanks. This was a regression from a bad conflict resolution while rebasing the PR.

File: src/providers/postgres/qgspostgresprovider.cpp (hunk 10 above)
Comment (ptitjano) [reply src/providers/postgres/qgspostgresprovider.cpp]: Applied. Thanks.

File: tests/src/python/test_qgsdelimitedtextprovider.py (hunk 12 above)
Comment (ptitjano) [reply tests/src/python/test_qgsdelimitedtextprovider.py]: Indeed. I have switched to `self.assertAlmostEqual`  on each coordinate in the meantime. This is how the other providers test the extent.

File: src/providers/postgres/raster/qgspostgresrasterprovider.h [hunk 13]
```diff
@@ -126,9 +126,9 @@ class QgsPostgresRasterProvider : public QgsRasterDataProvider
     //! Has spatial index
     bool mHasSpatialIndex = false;
     //! Raster size x
-    qlonglong mWidth = 0;
+    long mWidth = 0;
     //! Raster size y
-    qlonglong mHeight = 0;
+    long mHeight = 0;
```
Comment (nyalldawson) [inline src/providers/postgres/raster/qgspostgresrasterprovider.h:131]: why this change?

File: src/providers/postgres/qgspostgresprovider.cpp [hunk 14]
```diff
@@ -2212,7 +2227,11 @@ bool QgsPostgresProvider::parseDomainCheckConstraint( QStringList &enumValues, c
       if ( arrayPosition != -1 )
       {
         QString valueList = checkDefinition.mid( arrayPosition + 6, closingBracketPos );
+#if QT_VERSION < QT_VERSION_CHECK(5, 15, 0)
+        const QStringList commaSeparation = valueList.split( ',', QString::SkipEmptyParts );
+#else
```
Comment (nyalldawson) [inline src/providers/postgres/qgspostgresprovider.cpp:2232]: this can be dropped, we require 5.15 now

File: src/core/providers/ogr/qgsogrproviderutils.cpp [hunk 15]
```diff
@@ -3106,6 +3106,92 @@ OGRErr QgsOgrLayer::GetExtent( OGREnvelope *psExtent, bool bForce )
...
+        }
+        ResetReading();
+      }
+    }
+  }
+
+  if ( err != OGRERR_NONE )
+  {
+    QgsDebugMsgLevel( QStringLiteral( "Will computeExtent3DSlowly!" ), 3 );
+    err = computeExtent3DSlowly( psExtent3D );
```
Comment (nyalldawson) [inline src/core/providers/ogr/qgsogrproviderutils.cpp:3156]: Don't we still have a performance regression here? 

Ie in the situation:

- a driver which doesn't support fast 3d extent retrieval from OGR_L_GetExtent3D OR on older GDAL and the SELECT ST_MinZ... fails. 
- dataset has z
- the driver DOES support FAST retrieval of 2 extent
- ONLY the api to retrieve the 2d extent is called (eg to zoom to layer extent on loading the layer)

In that situation we're now forcing a complete table traversal when asking the provider for the 2d extent, because the 3d extent is calculated first. Whereas before we could quickly retrieve the 2d extent without table traversal...

File: src/providers/postgres/qgspostgresprovider.cpp (hunk 10 above)
Comment (benoitdm-oslandia) [reply src/providers/postgres/qgspostgresprovider.cpp]: > Z?: Add a "?" here. Because, without "?", you only look for "Z" type. But, maybe it's normal?

Yes it was intentional!

File: src/providers/postgres/raster/qgspostgresrasterprovider.h (hunk 13 above)
Comment (benoitdm-oslandia) [reply src/providers/postgres/raster/qgspostgresrasterprovider.h:131]: linked to clang check

File: src/providers/postgres/qgspostgresprovider.cpp (hunk 14 above)
Comment (benoitdm-oslandia) [reply src/providers/postgres/qgspostgresprovider.cpp:2232]: indeed

File: src/core/providers/ogr/qgsogrproviderutils.cpp (hunk 15 above)
Comment (benoitdm-oslandia) [reply src/core/providers/ogr/qgsogrproviderutils.cpp:3156]: hi @nyalldawson,

if the layer has z data and we have the 2D fast retrieval process for this driver, we still have to retrieve the minZ and maxZ data by using something like `SELECT ST_min(Z), ST_max(Z) FROM table`, which scan all rows. Also, by default for gdal 3.9, if the driver does not have an OGR_L_GetExtent3D implementation, gdal will scan all rows to compute the data (this is the same for the 2D extent).

Even if we use a second field to store the 2d extent when only the 2d extent is asked for a 3d layer, at some point, the 3D extent could be asked and we will have to scan all rows to compute it.

@rouault what do you think?

Comment (rouault) [reply src/core/providers/ogr/qgsogrproviderutils.cpp:3156]: > @rouault what do you think?

Will Extent3D will be called when adding a new vector layer ? If so, that might break some use cases. For example, currently you can add ``/vsizip//vsicurl/https://data.3dbag.nl/gpkg/v20231008/3dbag_nl.gpkg.zip`` (cf https://3dbag.nl/en/download . 19 GB as a ZIP) and use it reasonably with QGIS. But if Extent3D() is asked on it, then this will stall QGIS for hours.

Comment (nyalldawson) [reply src/core/providers/ogr/qgsogrproviderutils.cpp:3156]: > Will Extent3D will be called when adding a new vector layer 

Yes it will, because `QgsOgrProvider::extent()` just calls `return extent3D().toRectangle();`, and the layer extent is always required when adding a layer to a project (eg to zoom to the first layer added's extent in a new project)

Comment (nyalldawson) [reply src/core/providers/ogr/qgsogrproviderutils.cpp:3156]: Pragmatically I think we should leave QgsOgrProvider::extent() untouched, and ensure that we NEVER ask for the 3D extent in usual operations (ie not when adding layers or opening the layer properties windows).

Comment (rouault) [reply src/core/providers/ogr/qgsogrproviderutils.cpp:3156]: > Pragmatically I think we should leave QgsOgrProvider::extent() untouched, and ensure that we NEVER ask for the 3D extent in usual operations

I also concur with this

Comment (benoitdm-oslandia) [reply src/core/providers/ogr/qgsogrproviderutils.cpp:3156]: I have updated the PR to:
* use the old extent 2D code
* use 2 different field members to save the 2D extent and the 3D extent

The behaviour is now what you expected: we only load 3D extent when the 3D extent is asked.

Furthermore we have found a strange behaviour: 
a geopackage which contains 3D data but according to the metadata, the geometry is 2D. For example, the `lod12_2D` layer from the `dbag_nl.gpkg.zip file.   
In that case, the 3D extent will never be computed because the provider reads the extent information from the metadata which says this is a 2D layer.

This behaviour is not induced by this PR but just "revealed".

Comment (rouault) [reply src/core/providers/ogr/qgsogrproviderutils.cpp:3156]: > I have updated the PR to:
> 
>     * use the old extent 2D code
> 
>     * use 2 different field members to save the 2D extent and the 3D extent

LGTM !

> Furthermore we have found a strange behaviour:
> a geopackage which contains 3D data but according to the metadata, the geometry is 2D. For example, the `lod12_2D` layer from the `dbag_nl.gpkg.zip file.
> In that case, the 3D extent will never be computed because the provider reads the extent information from the metadata which says this is a 2D layer.

Ok, that's interesting. That lod12_2D layer doesn't set the Z=1 or 2 hint in the gpkg_geometry_columns, which is used by the driver to infer the dimensionality of the layer geometry type. I've fixed the writing side of the driver to set Z=2 (may contain 3D geoms) when one is encountered in https://github.com/OSGeo/gdal/pull/9147 . That said, it is a bit weird that the data producer generates a layre called 2D but which contains only 3D geometries...

File: src/core/vector/qgsvectorlayer.cpp [hunk 16]
```diff
@@ -1018,6 +1018,12 @@ QgsRectangle QgsVectorLayer::extent() const
 {
   QGIS_PROTECT_QOBJECT_THREAD_ACCESS
 
+  // if data is 3D, redirect to 3D extend computation, and save it as 3D extent
+  if ( mDataProvider && mDataProvider->elevationProperties() && mDataProvider->elevationProperties()->containsElevationData() )
+  {
+    return extent3D().toRectangle();
+  }
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayer.cpp:1025]: This change here should be revert too right, otherwise we're again potentially calculating an expensive 3d extent when we only need the 2d extent.

**Comment by nyalldawson** [comment]
Nearly there :+1:

File: src/core/providers/ogr/qgsogrproviderutils.cpp (hunk 15 above)
Comment (benoitdm-oslandia) [reply src/core/providers/ogr/qgsogrproviderutils.cpp:3156]: > That said, it is a bit weird that the data producer generates a layre called 2D but which contains only 3D geometries...

indeed...

File: src/core/vector/qgsvectorlayer.cpp (hunk 16 above)
Comment (benoitdm-oslandia) [reply src/core/vector/qgsvectorlayer.cpp]: my bad!

**Comment by benoitdm-oslandia** [comment]
fix tests to handle 2d/3d extent changes. 

@rouault There is something strange with the `test_provider_ogr_sqlite.py` test: it seems the `ST_MinZ` and `ST_MaxZ` do not work: they always return 0.0 . Is this expected?

**Comment by benoitdm-oslandia** [comment]
@rouault have you an idea why `testExtentSqlite` https://github.com/qgis/QGIS/blob/3b93870b48d5bd72b677b68c52a1d2cc12d33900/tests/src/python/test_provider_ogr_sqlite.py#L475 fails at line 566? 

With `QGIS_DEBUG=3` I saw the function `QgsOgrLayer::GetExtent3D` prints `done with Z! 0.0/0.0` at line https://github.com/qgis/QGIS/blob/3b93870b48d5bd72b677b68c52a1d2cc12d33900/src/core/providers/ogr/qgsogrproviderutils.cpp#L3146

Regards

**Comment by rouault** [comment]
> have you an idea why `testExtentSqlite`

yes, I gave you the reason in previous comment ;-)

Change
``ds = ogr.GetDriverByName('SQLite').CreateDataSource(tmpfile)``
to
``ds = ogr.GetDriverByName('SQLite').CreateDataSource(tmpfile, options=["SPATIALITE=YES"])``

But this probably means you should be more careful in the OGR providerl when using ST_MinZ/ST_MaxZ for a SQLite DB. You should like issue a ExecuteSQL("SELECT spatialite_version()") first and see if it returns something or an error

**Comment by rouault** [comment]
Merging.

**Comment by benoitdm-oslandia** [comment]
Thanks @nyalldawson @rouault for the review process!
